### PR TITLE
[#29697] Add prism artifact building workflow.

### LINKED ISSUE #29697: [Task] Publish Prism binary (os + architecture combos) artifacts on Beam Release.
Eventually we should publish OS + Architecture specific binaries for Prism in order to make it easier to use with released versions, and avoid requiring users compile it locally with Go (if possible). Though, the "build locally" solution can always remain a fallback, in particular for dev use, or novel OS + Arch combinations we don't release.

Maintained Actions exist: https://github.com/marketplace/actions/go-release-binaries  (would need to be INFRA approved, if not already) (More investigation into alternatives is recommended).

Otherwise, programmatic access exists for uploading to a release, and then downloading release assets:  https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#upload-a-release-asset

https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#get-a-release-asset

There's no limit/quotas associate with github assets, outside of the 2GB per file: https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases#storage-and-bandwidth-quotas

Then ultimately incorporating it into the release workflow for use with RCs and then the final releases.

### PR DESCRIPTION
**Pull request by lostluck**
Adds a build & stage to RC creation to release signed hashed and zipped prism binaries to a release.

Currently requires a release with the RC tag to exist, actually straightening out the logic is best handled in a separate PR, as this change is sufficiently complex as is without including the clobber protection logic if targetting an already published release.

Changes the workflow dispatch to us a JSON object instead of individual string fields. The dialog now looks like this:

image

Not an ideal experience, but we hit the 10 input limit otherwise.

The action syncs down the Dev Apache SVN repo, builds prism binaries for ARM64, AMD64 for Linux, Darwin (osx), and Windows, zipping up the binary, creating signature and sha512 hash files.

These are uploaded to a Github prerelease for the RC tag, like so: https://github.com/apache/beam/releases/tag/untagged-819e4037252d95b05550

Finally, the same artifacts are committed to the Dev Apache repository for final publishing.

Signature and hash are not included in the Zip so the archive doesn't need to be decompressed before being able to validate it.

TODO in a subsequent PR:
* automatically create a draft prerelease release for the tag if one doesn't exist yet.
* Update existing draft GH release for the release version with the new RC tag.
* Fail early if a published release has already gone out for the version to avoid clobbering an existing non-RC GH release.

* Updating the Release Manager guide.

Part of #29697

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
thanks. The refactoring makes sense as there are more and more fields in the action which is inconvenient. Just wondering if there is a way to keep the descriptions of field concise. Currently the dropdown is very long. (It's fine to keep as is for now)

File: .github/workflows/build_release_candidate.yml
Comment (Abacn) [inline .github/workflows/build_release_candidate.yml:493]: Do we need a full matrix of 2x3=6 builds. Fine to leave as is but just wondering if cross-compiled archs takes take long time to  build

**Review by Abacn — COMMENTED** [review]
thanks, had two optional comments.

Do we plan to deliver prism binaries for Beam 2.57.0 (release cut soon)

File: .github/workflows/build_release_candidate.yml
Comment (lostluck) [reply .github/workflows/build_release_candidate.yml:493]: The entire action takes ~3 minutes total, so the additional architectures don't currently add much time. If they did, we'd move to a set up closer to the Python wheels

**Comment by lostluck** [comment]
> thanks, had two optional comments.
> 
> Do we plan to deliver prism binaries for Beam 2.57.0 (release cut soon)

Yup. That's why I was working on this all last week, to figure out pand solve problems before we ran into them in the release.

It's easier to get the SDK downloading step working with a proper release having the artifacts. That half is for 2.58.0 though, and not 2.57.0.

**Comment by lostluck** [comment]
> thanks. The refactoring makes sense as there are more and more fields in the action which is inconvenient. Just wondering if there is a way to keep the descriptions of field concise. Currently the dropdown is very long. (It's fine to keep as is for now)

Unfortunately I tried a few things for this and nothing made it better. We're ultimately going to need to keep the release guide documentation (which is in a later PR), paired with this and link to the documentation here to reduce the visual noise.

TBH, we'd probably move the initiation for this action to a cooy/pastable command line to do the workflow dispatch. See https://cli.github.com/manual/gh_workflow_run

**Review by Abacn — APPROVED** [review]
If Want LGTM from the release manager then it's good to add milestone to this PR. Otherwise this LGTM and feel free to merge

**Comment by lostluck** [comment]
Since Kenn's going to affected by it, hopefully he'll review it first!

**Comment by lostluck** [comment]
Merging in, but I'm still going to be working with @kennknowles to make sure this is iterated on for completeness.

**Pull request merged**
